File: android-sdk-ui/src/main/java/com/braze/ui/inappmessage/views/InAppMessageBoundedLayout.java
package com.braze.ui.inappmessage.views;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.widget.RelativeLayout;

import com.appboy.ui.R;

/**
 * A {@link RelativeLayout} that respects maximum/minimum dimension bounds.
 */
public class InAppMessageBoundedLayout extends RelativeLayout {
  private int mMaxDefinedWidthPixels;
  private int mMinDefinedWidthPixels;
  private int mMaxDefinedHeightPixels;
  private int mMinDefinedHeightPixels;

  public InAppMessageBoundedLayout(Context context) {
    super(context);
  }

  public InAppMessageBoundedLayout(Context context, AttributeSet attrs) {
    super(context, attrs);
    TypedArray attributes = context.obtainStyledAttributes(attrs, R.styleable.InAppMessageBoundedLayout);
    mMaxDefinedWidthPixels = attributes.getDimensionPixelSize(R.styleable.InAppMessageBoundedLayout_inAppMessageBoundedLayoutMaxWidth, 0);
    mMinDefinedWidthPixels = attributes.getDimensionPixelSize(R.styleable.InAppMessageBoundedLayout_inAppMessageBoundedLayoutMinWidth, 0);
    mMaxDefinedHeightPixels = attributes.getDimensionPixelSize(R.styleable.InAppMessageBoundedLayout_inAppMessageBoundedLayoutMaxHeight, 0);
    mMinDefinedHeightPixels = attributes.getDimensionPixelSize(R.styleable.InAppMessageBoundedLayout_inAppMessageBoundedLayoutMinHeight, 0);
    attributes.recycle();
  }

  @Override
  protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
    int measuredWidth = MeasureSpec.getSize(widthMeasureSpec);
    if (mMinDefinedWidthPixels > 0 && measuredWidth < mMinDefinedWidthPixels) {
      int measureMode = MeasureSpec.getMode(widthMeasureSpec);
      widthMeasureSpec = MeasureSpec.makeMeasureSpec(mMinDefinedWidthPixels, measureMode);
    } else if (mMaxDefinedWidthPixels > 0 && measuredWidth > mMaxDefinedWidthPixels) {
      int measureMode = MeasureSpec.getMode(widthMeasureSpec);
      widthMeasureSpec = MeasureSpec.makeMeasureSpec(mMaxDefinedWidthPixels, measureMode);
    }

    int measuredHeight = MeasureSpec.getSize(heightMeasureSpec);
    if (mMinDefinedHeightPixels > 0 && measuredHeight < mMinDefinedHeightPixels) {
      int measureMode = MeasureSpec.getMode(heightMeasureSpec);
      heightMeasureSpec = MeasureSpec.makeMeasureSpec(mMinDefinedHeightPixels, measureMode);
    } else if (mMaxDefinedHeightPixels > 0 && measuredHeight > mMaxDefinedHeightPixels) {
      int measureMode = MeasureSpec.getMode(heightMeasureSpec);
      heightMeasureSpec = MeasureSpec.makeMeasureSpec(mMaxDefinedHeightPixels, measureMode);
    }
    super.onMeasure(widthMeasureSpec, heightMeasureSpec);
  }
}

File: android-sdk-ui/src/main/java/com/braze/ui/inappmessage/views/InAppMessageBoundedLayout.kt
package com.braze.ui.inappmessage.views

import android.content.Context
import android.util.AttributeSet
import android.widget.RelativeLayout
import com.appboy.ui.R

/**
 * A [RelativeLayout] that respects maximum/minimum dimension bounds.
 */
open class InAppMessageBoundedLayout : RelativeLayout {
    private var maxDefinedWidthPixels = 0
    private var minDefinedWidthPixels = 0
    private var maxDefinedHeightPixels = 0
    private var minDefinedHeightPixels = 0

    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        val attributes =
            context.obtainStyledAttributes(attrs, R.styleable.InAppMessageBoundedLayout)
        maxDefinedWidthPixels = attributes.getDimensionPixelSize(
            R.styleable.InAppMessageBoundedLayout_inAppMessageBoundedLayoutMaxWidth,
            0
        )
        minDefinedWidthPixels = attributes.getDimensionPixelSize(
            R.styleable.InAppMessageBoundedLayout_inAppMessageBoundedLayoutMinWidth,
            0
        )
        maxDefinedHeightPixels = attributes.getDimensionPixelSize(
            R.styleable.InAppMessageBoundedLayout_inAppMessageBoundedLayoutMaxHeight,
            0
        )
        minDefinedHeightPixels = attributes.getDimensionPixelSize(
            R.styleable.InAppMessageBoundedLayout_inAppMessageBoundedLayoutMinHeight,
            0
        )
        attributes.recycle()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        var newWidthMeasureSpec = widthMeasureSpec
        var newHeightMeasureSpec = heightMeasureSpec

        val measuredWidth = MeasureSpec.getSize(newWidthMeasureSpec)
        if (minDefinedWidthPixels > 0 && measuredWidth < minDefinedWidthPixels) {
            val measureMode = MeasureSpec.getMode(newWidthMeasureSpec)
            newWidthMeasureSpec = MeasureSpec.makeMeasureSpec(minDefinedWidthPixels, measureMode)
        } else if (maxDefinedWidthPixels > 0 && measuredWidth > maxDefinedWidthPixels) {
            val measureMode = MeasureSpec.getMode(newWidthMeasureSpec)
            newWidthMeasureSpec = MeasureSpec.makeMeasureSpec(maxDefinedWidthPixels, measureMode)
        }

        val measuredHeight = MeasureSpec.getSize(newHeightMeasureSpec)
        if (minDefinedHeightPixels > 0 && measuredHeight < minDefinedHeightPixels) {
            val measureMode = MeasureSpec.getMode(newHeightMeasureSpec)
            newHeightMeasureSpec = MeasureSpec.makeMeasureSpec(minDefinedHeightPixels, measureMode)
        } else if (maxDefinedHeightPixels > 0 && measuredHeight > maxDefinedHeightPixels) {
            val measureMode = MeasureSpec.getMode(newHeightMeasureSpec)
            newHeightMeasureSpec = MeasureSpec.makeMeasureSpec(maxDefinedHeightPixels, measureMode)
        }
        super.onMeasure(newWidthMeasureSpec, newHeightMeasureSpec)
    }
}
